### ISSUE TITLE
NaN with modified attention

### ISSUE THREAD

**Issue by sebastien-j** [kyunghyuncho/dl4mt-material#31]
@laulysta, @jakezhaojb, @kyunghyuncho 

This might be related to issue #29. I'm having numerical issues (NaNs) with the modified attention introduced in pull request #15.

I'll go though the code to see what it does soon. What are the equations that are supposed to describe the modified GRU with attention? Do you have empirical results comparing both versions?

**Comment by orhanf** [comment]
@sebastien-j i dont think this issue is related with #29 (that is just a problem for out-of-sync `nmt.py` after pr #15 - i've synced them [here](https://github.com/orhanf/dl4mt-material/commit/9199ac05717f37b13694c4c8ce9d5d0938d746c6) but not yet made a pr). 
Not in this repo, but i've adapted this modified GRU to multi-way translation and it was performing slightly better. From what i've experimented, NaNs arise with Adam having a learning rate bigger than 1e-3.

**Comment by sebastien-j** [comment]
Thanks @orhanf.

I'm using adadelta. After all, the NaN issue may not be related to the different GRU variant as it just crashed with the previous version. I'll try different datasets/preprocessing to try to understand what is going on.

**Comment by orhanf** [comment]
One more thing that comes to my mind is stabilizing the attention softmax, [this](https://github.com/orhanf/dl4mt-material/commit/b0f062738a925541d967cedb63a44d07a570b1b9) enhancement. We were getting NaNs because of that instability (in blocks implementation tho)

**Comment by sebastien-j** [comment]
This is strange. Out of 2 computers I tried, the bug only occurs on one. The Theano versions are identical, and I added some code to allow choosing a RNG.

In any case, it appears that the problem is not directly related to the dl4mt code.

**Comment by amirj** [comment]
I have a same problem with NANs. Using smaller batch_size(=32) solved it.
